core/lomas_core/models/responses.py

```diff
@@ -0,0 +1,170 @@
+from typing import Annotated, Dict, List, Literal, Union
+
+import pandas as pd
+from diffprivlib.validation import DiffprivlibMixin
+from pydantic import (
+    BaseModel,
+    ConfigDict,
+    Discriminator,
+    PlainSerializer,
+    PlainValidator,
+    ValidationInfo,
+    field_validator,
+)
+from snsynth import Synthesizer
+
+from lomas_core.models.utils import (
+    dataframe_from_dict,
+    dataframe_to_dict,
+    deserialize_model,
+    serialize_model,
+)
+
+
+class ResponseModel(BaseModel):
+    """Base model for any response from the server."""
+
+
+class InitialBudgetResponse(ResponseModel):
+    """Model for responses to initial budget queries."""
+
+    initial_epsilon: float
+    initial_delta: float
+
+
+class SpentBudgetResponse(ResponseModel):
+    """Model for responses to spent budget queries."""
+
+    total_spent_epsilon: float
+    total_spent_delta: float
+
+
+class RemainingBudgetResponse(ResponseModel):
+    """Model for responses to remaining budget queries."""
+
+    remaining_epsilon: float
+    remaining_delta: float
+
+
+class DummyDsResponse(ResponseModel):
+    """Model for responses to dummy dataset requests."""
+
+    model_config = ConfigDict(arbitrary_types_allowed=True)
+    dtypes: Dict[str, str]
+    datetime_columns: List[str]
+    dummy_df: Annotated[pd.DataFrame, PlainSerializer(dataframe_to_dict)]
+
+    @field_validator("dummy_df", mode="before")
+    @classmethod
+    def deserialize_dummy_df(
+        cls, v: pd.DataFrame | dict, info: ValidationInfo
+    ) -> pd.DataFrame:
+        """Decodes the dict representation of the dummy df with correct types.
+
+        Only does so if the input value is not already a dataframe.
+        Args:
+            v (pd.DataFrame | dict): The dataframe to decode.
+            info (ValidationInfo): Validation info to access other model fields.
+
+        Returns:
+            pd.DataFrame: The decoded dataframe.
+        """
+        if isinstance(v, pd.DataFrame):
+            return v
+
+        dtypes = info.data["dtypes"]
+        datetime_columns = info.data["datetime_columns"]
+        dummy_df = dataframe_from_dict(v)
+        dummy_df = dummy_df.astype(dtypes)
+        for col in datetime_columns:
+            dummy_df[col] = pd.to_datetime(dummy_df[col])
+        return dummy_df
+
+
+class CostResponse(ResponseModel):
+    """Model for responses to cost estimation requests."""
+
+    epsilon: float
+    delta: float
+
+
+# Query Responses
+# -----------------------------------------------------------------------------
+
+
+# DiffPrivLib
+class DiffPrivLibQueryResult(BaseModel):
+    """Model for diffprivlib query result."""
+
+    model_config = ConfigDict(arbitrary_types_allowed=True)
+    res_type: Literal["diffprivlib"] = "diffprivlib"
```

Comment:
I think it would be nice if the res_type Litterals of the query responses models were replaced with the DPLibraries StrEnum from lomas_core/constants.
Like DPLibraries.DIFFPRIVLIB instead of "diffprivlib"

Reply:
thanks! will be changed in next commit